### PR TITLE
Thread-safe access to EntityTree in OverlayPanel

### PR DESCRIPTION
**Pull request by AlessandroSigna**
Binding an overlay to an entity caused crashes. This PR seems to fix the problem

### CODE REVIEW

File: interface/src/ui/overlays/OverlayPanel.cpp [hunk 1]
```diff
@@ -170,9 +170,11 @@ void OverlayPanel::applyTransformTo(Transform& transform, bool force) {
                 transform.setTranslation(DependencyManager::get<AvatarManager>()->getMyAvatar()
                                          ->getPosition());
             } else if (!_anchorPositionBindEntity.isNull()) {
-                transform.setTranslation(DependencyManager::get<EntityScriptingInterface>()
-                                         ->getEntityTree()->findEntityByID(_anchorPositionBindEntity)
-                                         ->getPosition());
+                EntityTreePointer entityTree = DependencyManager::get<EntityScriptingInterface>()->getEntityTree();
+                entityTree = DependencyManager::get<EntityScriptingInterface>()->getEntityTree();
```
Comment (huffman) [inline interface/src/ui/overlays/OverlayPanel.cpp]: This line seems to be a duplicate of the above line?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 57bda9d5de9d0fe675c70688e2e03c78a3ffb119 available at: https://gist.github.com/c0b147fbd033c707e475<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6786.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6786.dmg

File: interface/src/ui/overlays/OverlayPanel.cpp (hunk 1 above)
Comment (AlessandroSigna) [inline interface/src/ui/overlays/OverlayPanel.cpp]: Ops, yea

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3cf37ea31d00c16fd68e903fa6f165b860b2e18e available at: https://gist.github.com/c0b147fbd033c707e475<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6786.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6786.dmg

File: interface/src/ui/overlays/OverlayPanel.cpp [hunk 2]
```diff
@@ -170,9 +170,10 @@ void OverlayPanel::applyTransformTo(Transform& transform, bool force) {
                 transform.setTranslation(DependencyManager::get<AvatarManager>()->getMyAvatar()
                                          ->getPosition());
             } else if (!_anchorPositionBindEntity.isNull()) {
-                transform.setTranslation(DependencyManager::get<EntityScriptingInterface>()
-                                         ->getEntityTree()->findEntityByID(_anchorPositionBindEntity)
-                                         ->getPosition());
+                EntityTreePointer entityTree = DependencyManager::get<EntityScriptingInterface>()->getEntityTree();
+                entityTree->withReadLock([&] {
+                    transform.setTranslation(entityTree->findEntityByID(_anchorPositionBindEntity)->getPosition());
```
Comment (ZappoMan) [inline interface/src/ui/overlays/OverlayPanel.cpp]: this is not safe, findEntityByID() can return NULL and that will crash on dereferencing `getPosition()`

File: interface/src/ui/overlays/OverlayPanel.cpp [hunk 3]
```diff
@@ -181,9 +182,10 @@ void OverlayPanel::applyTransformTo(Transform& transform, bool force) {
                 transform.setRotation(DependencyManager::get<AvatarManager>()->getMyAvatar()
                                       ->getOrientation());
             } else if (!_anchorRotationBindEntity.isNull()) {
-                transform.setRotation(DependencyManager::get<EntityScriptingInterface>()
-                                      ->getEntityTree()->findEntityByID(_anchorRotationBindEntity)
-                                      ->getRotation());
+                EntityTreePointer entityTree = DependencyManager::get<EntityScriptingInterface>()->getEntityTree();
+                entityTree->withReadLock([&] {
+                    transform.setRotation(entityTree->findEntityByID(_anchorRotationBindEntity)->getRotation());
```
Comment (ZappoMan) [inline interface/src/ui/overlays/OverlayPanel.cpp]: same here

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7fc921291ffecc48210ebd5518c190b6ed8ee3d3 available at: https://gist.github.com/c0b147fbd033c707e475<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6786.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6786.dmg

File: interface/src/ui/overlays/OverlayPanel.cpp [hunk 4]
```diff
@@ -170,9 +170,13 @@ void OverlayPanel::applyTransformTo(Transform& transform, bool force) {
                 transform.setTranslation(DependencyManager::get<AvatarManager>()->getMyAvatar()
                                          ->getPosition());
             } else if (!_anchorPositionBindEntity.isNull()) {
-                transform.setTranslation(DependencyManager::get<EntityScriptingInterface>()
-                                         ->getEntityTree()->findEntityByID(_anchorPositionBindEntity)
-                                         ->getPosition());
+                EntityTreePointer entityTree = DependencyManager::get<EntityScriptingInterface>()->getEntityTree();
+                entityTree->withReadLock([&] {
+                    EntityItemPointer foundEntity = entityTree->findEntityByID(_anchorPositionBindEntity);
+                    if (foundEntity != NULL) {
```
Comment (ZappoMan) [inline interface/src/ui/overlays/OverlayPanel.cpp]: this is better... but the coding standard says this should be.... `if (foundEntity) {` no need to compare to null

**Comment by hifi-gustavo** [comment]
Build console output for commit: eaa67a41dd6e62cb739919ec6a938089f57be862 available at: https://gist.github.com/c0b147fbd033c707e475<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6786.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6786.dmg

**Comment by ZappoMan** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: eaa67a41dd6e62cb739919ec6a938089f57be862 available at: https://gist.github.com/c0b147fbd033c707e475<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6786.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6786.dmg

**Pull request merged**